api/src/main/java/io/cloudevents/Extension.java

```diff
@@ -17,12 +17,16 @@
 
 package io.cloudevents;
 
-import java.util.Map;
-
-public interface Extension {
-
-    void readFromEvent(CloudEvent event);
+/**
+ * Materialized CloudEvent Extension interface to read/write the extension attributes key/values.
+ */
+public interface Extension extends CloudEventExtensions {
```

Comment:
This sounds wrong. `Extension extends CloudEventExtensions` smells.

Especially given that it will look like:
```java
Extension ext = ...;

ext.getExtension("OMG WTF!");
ext.getExtensionNames(); // whaaaat?
```

Reply:
how about:

```
public interface Extension {

    void readFrom(CloudEventExtensions extensions);

   Object getValue(String key);

   Set<String> getKeys();

}
```

Reply:
Much better :)

WDYT about `Iterable<String> getKeys();`? Since extension may define a dynamic set of keys, `Set` here can be a bit restrictive, and it does not look like we need to have `Set`'s characteristics in this case

Reply:
To me it doesn't sound explicative `Iterable<String>`, because `Iterable` can contain duplicates. Because we're returning a key set, to me it makes sense using the `Set<String>` type, exactly like i expect a Map to return the keys as a set

Reply:
On the other hand, I don't think it would be a bad idea to let this class extend `Iterable<Map.Entry<String, Object>>` or have a method like `iter()` that returns such `Iterable`

Reply:
And btw such `Iterable` can be default implemented

Reply:
`Iterable<Map.Entry<String, Object>>` would require allocations.

Okay, let's keep `Set` then.

Reply:
> Iterable<Map.Entry<String, Object>> would require allocations.

Sure, it's like "use it at your own risk" :smile: